libs/partners/google-genai/langchain_google_genai/chat_models.py

```diff
@@ -332,14 +333,90 @@ def _parse_chat_history(
     return messages
 
 
-def _parts_to_content(parts: List[genai.types.PartType]) -> Union[List[dict], str]:
+def _retrieve_function_call_response(
+    parts: List[genai.types.PartType],
+) -> Optional[Dict]:
+    for idx, part in enumerate(parts):
+        if part.function_call and part.function_call.name:
+            fc: FunctionCall = part.function_call
+            return {
+                "function_call": {
+                    "name": fc.name,
+                    "arguments": dict(fc.args.items()),
+                }
+            }
+    return None
+
+
+def _convert_function_call_req(function_calls: Union[Dict, List[Dict]]) -> Dict:
+    function_declarations = []
+    if isinstance(function_calls, dict):
+        function_declarations.append(_convert_fc_type(function_calls))
+    else:
+        for fc in function_calls:
+            function_declarations.append(_convert_fc_type(fc))
+    return {
+        "function_declarations": function_declarations,
+    }
+
+
+def _convert_fc_type(fc: Dict) -> Dict:
+    # type_: "Type"
+    # format_: str
+    # description: str
+    # nullable: bool
+    # enum: MutableSequence[str]
+    # items: "Schema"
+    # properties: MutableMapping[str, "Schema"]
+    # required: MutableSequence[str]
+    if "parameters" in fc:
+        fc["parameters"] = _convert_fc_type(fc["parameters"])
+    if "properties" in fc:
+        for k, v in fc["properties"].items():
+            fc["properties"][k] = _convert_fc_type(v)
+    if "type" in fc:
+        # STRING = 1
+        # NUMBER = 2
+        # INTEGER = 3
+        # BOOLEAN = 4
+        # ARRAY = 5
+        # OBJECT = 6
+        if fc["type"] == "string":
+            fc["type_"] = 1
+        elif fc["type"] == "number":
+            fc["type_"] = 2
+        elif fc["type"] == "integer":
+            fc["type_"] = 3
+        elif fc["type"] == "boolean":
+            fc["type_"] = 4
+        elif fc["type"] == "array":
+            fc["type_"] = 5
+        elif fc["type"] == "object":
+            fc["type_"] = 6
+        del fc["type"]
+    if "format" in fc:
+        fc["format_"] = fc["format"]
+        del fc["format"]
+
+    for k, v in fc.items():
+        if isinstance(v, dict):
+            fc[k] = _convert_fc_type(v)
+
+    return fc
+
+
+def _parts_to_content(
```

Comment:
what if we simplify it, and parse directly to a message?

Reply:
for future followup